test/Parse/enum_element_pattern_swift4.swift

```diff
@@ -0,0 +1,21 @@
+// RUN: %target-typecheck-verify-swift -swift-version 4
+
+// https://bugs.swift.org/browse/SR-3452
+// See test/Compatibility/enum_element_pattern.swift for Swift3 behavior.
+
+enum E {
+  case A, B, C
+}
+
+func testE(e: E) {
+  switch e {
+  case E.A<UndefinedTy>(): // expected-error {{use of undeclared type 'UndefinedTy'}}
+    break
+  case E.B<Int>(): // expected-error {{cannot specialize a non-generic definition}} expected-note {{while parsing this '<' as a type parameter bracket}}
+    break
+  case E.C(): // Ok.
```

Comment:
Thanks!
Unfortunately, this compiles in Swift3.  
```swift
enum E {
  case A, B, C
}

func testE(e: E) {
  switch e {
  case E.C(let payload):
    print(payload) // prints "()"
  default:
    break
  }
}
```
For this case, I think we can't drop `(let payload)` from AST in Swift3 mode.

```swift
enum E {
  case A, B, C
}

func testE(e: E) {
  guard case E.C(let payload) = e { return }
  print(payload)
}
```
does assert-fail in SILGen.
In no-assert Swift3 GM, IRGen (Verifer) emits:
```
Global is external, but doesn't have external or weak linkage!
i8* ()* @_TFF4test5testEFT1eOS_1E_T_auL_7payloadT_
invalid linkage type for function declaration
i8* ()* @_TFF4test5testEFT1eOS_1E_T_auL_7payloadT_
LLVM ERROR: Broken module found, compilation aborted!
```